tests/pubsub/testgossipsubheartbeat.nim

```diff
@@ -0,0 +1,305 @@
+    tryPublish await node0.publish(topic, newSeq[byte](10000)), 1
+
+    # Then Node0 fanout peers are populated
+    await waitForPeersInTable(node0, topic, 6, PeerTableType.Fanout)
+    check:
+      node0.fanout.hasKey(topic) and node0.fanout[topic].len == 6
+
+    # When peers of Node0 fanout are disconnected
+    let peersToDisconnect = node0.fanout[topic].toSeq()[1 .. ^1].mapIt(it.peerId)
+    await findAndStopPeers(nodes, peersToDisconnect, topic, voidTopicHandler)
+    await waitForPeersInTable(node0, topic, 1, PeerTableType.Fanout)
+
+    # Then Node0 fanout peers are replenished during heartbeat
+    await waitForHeartbeat()
+    check:
+      node0.fanout[topic].len == 4
+      node0.fanout[topic].toSeq().allIt(it.peerId notin peersToDisconnect)
+
+  asyncTest "iDontWants history - last element is pruned during heartbeat":
+    let
+      topic = "foobar"
+      nodes = generateNodes(
+          2, gossip = true, sendIDontWantOnPublish = true, historyLength = 5
+        )
+        .toGossipSub()
+
+    startNodesAndDeferStop(nodes)
+
+    await connectNodes(nodes[0], nodes[1])
+    subscribeAllNodes(nodes, topic, voidTopicHandler)
+    await waitForHeartbeat()
+
+    # When Node0 sends 10 messages to the topic
+    for i in 0 ..< 10:
+      tryPublish await nodes[0].publish(topic, newSeq[byte](1000)), 1
+
+    # Then Node1 receives 10 iDontWant messages from Node0
+    let peer = nodes[1].mesh[topic].toSeq()[0]
+    check:
+      peer.iDontWants[^1].len == 10
+
+    # When heartbeat happens
+    await waitForHeartbeat()
+
+    # Then last element of iDontWants history is pruned
+    check:
+      peer.iDontWants[^1].len == 0
+
+  asyncTest "sentIHaves history - last element is pruned during heartbeat":
+    # 3 Nodes, Node 0 <==> Node 1 and Node 0 <==> Node 2
+    # due to DValues: 1 peer in mesh and 1 peer only in gossip of Node 0
+    let
+      numberOfNodes = 3
+      topic = "foobar"
+      nodes = generateNodes(
+          numberOfNodes,
+          gossip = true,
+          historyLength = 3,
+          dValues =
+            some(DValues(dLow: some(1), dHigh: some(1), d: some(1), dOut: some(0))),
+        )
+        .toGossipSub()
+
+    startNodesAndDeferStop(nodes)
+
+    for i in 1 ..< numberOfNodes:
+      await connectNodes(nodes[0], nodes[i])
+    subscribeAllNodes(nodes, topic, voidTopicHandler)
+
+    # Waiting 2 heartbeats to populate sentIHaves history
+    await waitForHeartbeat(2)
+
+    # When Node0 sends a messages to the topic
+    for i in 0 ..< 1:
+      tryPublish await nodes[0].publish(topic, newSeq[byte](1000)), 1
+
+    # Find Peer outside of mesh to which Node 0 will send IHave
+    let peer =
+      nodes[0].gossipsub[topic].toSeq().filterIt(it notin nodes[0].mesh[topic])[0]
+
+    # When next heartbeat occurs
+    let timeout = Moment.now() + 100.milliseconds
+    while peer.sentIHaves[^1].len == 0:
+      await activeWait(1.milliseconds, timeout, "wait for sentIHaves timeout")
```

Comment:
1ms sleep interval seems vary small, depending on language, os, machine, and other factors, it's vary arguable if code will sleep (it may just skip). it would be better to use bit larger timeout like 10 ms.